File: leetcode/course-schedule/course-schedule-dfs-rec.rb
# @param {Integer} num_courses
# @param {Integer[][]} prerequisites
# @return {Boolean}
def can_finish(num_courses, prerequisites)
  !Graph.new(num_courses, prerequisites).cyclic?
end

class Graph
  def initialize(vertex_count, edges)
    @adj = Array.new(vertex_count) { [] }
    edges.each { |src, dest| @adj[src] << dest }
  end

  def cyclic?
    vis = [:white] * @adj.size

    cyclic_from = ->(src) do
      case vis[src]
      when :white
        vis[src] = :gray
        return true if @adj[src].any?(&cyclic_from)
        vis[src] = :black
        false
      when :gray
        true
      when :black
        false
      else
        raise 'Bug: unrecognized visitation state'
      end
    end

    (0...@adj.size).any?(&cyclic_from)
  end
end

File: leetcode/course-schedule/course-schedule-dfs-rec.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        var graph = new Graph(numCourses);
        for (var edge : prerequisites) graph.addEdge(edge[1], edge[0]);
        return !graph.hasCycle();
    }
}

class Graph {
    Graph(int vertexCount) {
        _adj = new ArrayList<>(vertexCount);
        while (_adj.size() < vertexCount) _adj.add(new ArrayList<>());
    }

    void addEdge(int src, int dest) {
        _adj.get(src).add(dest);
    }

    boolean hasCycle() {
        var vis = freshVisitationList();

        RecursiveIntPredicate hasCycleFrom = (me, src) -> {
            switch (vis[src]) {
            case WHITE:
                vis[src] = Color.GRAY;
                if (_adj.get(src).stream().anyMatch(dest -> me.test(me, dest)))
                    return true;
                vis[src] = Color.BLACK;
                return false;

            case GRAY:
                return true;

            case BLACK:
                return false;

            default:
                throw new AssertionError("unrecognized visitation state");
            }
        };

        return vertices()
                .anyMatch(start -> hasCycleFrom.test(hasCycleFrom, start));
    }

    private Color[] freshVisitationList() {
        return vertices().mapToObj(_vertex -> Color.WHITE)
                         .toArray(Color[]::new);
    }

    private IntStream vertices() {
        return IntStream.range(0, _adj.size());
    }

    private final List<List<Integer>> _adj;
}

enum Color {
    WHITE,
    GRAY,
    BLACK,
}

@FunctionalInterface
interface RecursiveIntPredicate {
    boolean test(RecursiveIntPredicate me, int value);
}
